### PR TITLE
ALL IS DONE

### PR DESCRIPTION
**Pull request by Sesue2002**
YEAH, im done.

### CODE REVIEW

File: src/main/java/Problem3/Book.java [hunk 1]
```diff
@@ -15,11 +15,24 @@ public Book(String title, String author) {
...
     public boolean equals(Object obj) {
-        // homework
+        if (this == obj){
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()){
+            return false;
+        }
+        Book b1 = (Book) obj;
+        return title.equals(b1.title) && author.equals(b1.author) && id == b1.id;
```
Comment (pdgetrf) [inline src/main/java/Problem3/Book.java:36]: wrong logic 

```
A movie equals another movie only when their id are the same. Same rule for books being equal.
```

File: src/main/java/Problem3/BookFiction.java [hunk 2]
```diff
@@ -9,12 +9,15 @@
...
         // tip: use the 'super' keyword
+        super(title, author);
+        this.genres = genres;
     }
 
     // copy constructor
     public BookFiction(BookFiction anotherBook) {
         // homework
         // tip: use the 'super' keyword
+        super(anotherBook);
```
Comment (pdgetrf) [inline src/main/java/Problem3/BookFiction.java:20]: missing genres

File: src/main/java/Problem3/Movie.java [hunk 3]
```diff
@@ -10,14 +10,31 @@
...
     public boolean equals(Object obj) {
         // homework
+        if (this == obj){
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()){
+            return false;
+        }
+        Movie m1 = (Movie) obj;
+        return rating.equals(m1.rating) && title.equals(m1.title) && id == m1.id;
```
Comment (pdgetrf) [inline src/main/java/Problem3/Movie.java:38]: same here.

File: src/main/java/Problem3/MovieAction.java [hunk 4]
```diff
@@ -22,5 +24,14 @@ public int getLateFeeInDollar() {
...
     public int calcLateFees(int numOfDaysPastDue) {
         // homework
+        if (numOfDaysPastDue <= 0){
+            return 0;
+        }
+        if(0 < numOfDaysPastDue && numOfDaysPastDue < 5){
+            return numOfDaysPastDue * 5;
+        }
+        else{
+            return 2 * (numOfDaysPastDue * 5);
```
Comment (pdgetrf) [inline src/main/java/Problem3/MovieAction.java:34]: duplicated logic here. 

suggesting:
```java
        int fee = lateFeePerDayInDollar * numOfDaysPastDue;
        return (numOfDaysPastDue >= 5) ? 2 * fee : fee;
```

and here number 5 only appears once too

File: src/test/java/Problem2Test.java [hunk 5]
```diff
@@ -7,13 +7,25 @@
...
         int[][] expects = {
-                {0, 1, 2, 4, 6}
+                {0, 1, 2, 4, 6},
                 // add more test cases
+                {0,1,2,3,4,5,6,7},
+                {0},
+                {0,1,2,3,4,5,6,7,8,9},
+                {0,1,2},
+                {0,1,2,3,4,5,6,7,8,9,10,11},
+                {0,2,3,6,7,8,9,10,12,15}
```
Comment (pdgetrf) [inline src/test/java/Problem2Test.java:28]: great test. let's add more cases for the edge cases, such as null array and empty array.

**Review by pdgetrf — APPROVED** [review]
LGTM
